src/main/java/de/tum/cit/aet/artemis/programming/service/GitService.java

```diff
@@ -1502,4 +1546,214 @@ private void getCommitInfo(Git git, List<CommitInfoDTO> commitInfos) throws GitA
+
+            // Step 1: Add the .git directory (full history)
+            Path bareRepoPath = repository.getDirectory().toPath();
+            addDirectoryToZip(zipOutputStream, bareRepoPath, bareRepoPath, ".git");
+
+            // Step 2: Add the working tree snapshot using ArchiveCommand
+            try {
+                ObjectId treeId = repository.resolve(treeish);
+                if (treeId != null) {
+                    try (ByteArrayOutputStream archiveStream = new ByteArrayOutputStream()) {
+                        try (Git git = new Git(repository)) {
+                            git.archive().setTree(treeId).setFormat("zip").setOutputStream(archiveStream).call();
+                        }
+
+                        // Extract the archive contents and add them to our main zip
+                        try (ZipInputStream zipInputStream = new ZipInputStream(new ByteArrayInputStream(archiveStream.toByteArray()))) {
+                            ZipEntry entry;
+                            while ((entry = zipInputStream.getNextEntry()) != null) {
+                                zipOutputStream.putNextEntry(new ZipEntry(entry.getName()));
+                                zipInputStream.transferTo(zipOutputStream);
+                                zipOutputStream.closeEntry();
+                            }
+                        }
+                    }
+                }
+                else {
+                    log.debug("Could not resolve tree for branch '{}', repository might be empty", branch);
+                }
+            }
+            catch (Exception e) {
+                log.debug("Could not create archive for branch '{}': {}", branch, e.getMessage());
+                // Continue without the working tree content, just include .git directory
+            }
+
+            zipOutputStream.finish();
+            byte[] zipData = outputStream.toByteArray();
+
+            return new InputStreamResource(new ByteArrayInputStream(zipData)) {
+
+                @Override
+                public String getFilename() {
+                    return filename + ".zip";
+                }
+
+                @Override
+                public long contentLength() {
+                    return zipData.length;
+                }
+            };
+        }
+    }
+
+    /**
+     * Helper method to add a directory and its contents to a ZIP output stream.
+     *
+     * @param zipOutputStream the ZIP output stream to write to
+     * @param rootPath        the root path for calculating relative paths
+     * @param pathToAdd       the path to add to the ZIP
+     * @param prefix          the prefix to use in the ZIP entry names
+     * @throws IOException if an I/O error occurs
+     */
+    private void addDirectoryToZip(ZipOutputStream zipOutputStream, Path rootPath, Path pathToAdd, String prefix) throws IOException {
+        Files.walk(pathToAdd).forEach(path -> {
```

Comment:
This can also throw a `SecurityException` if I am not mistaken, we should handle or throw it